src/MPS.jl

```diff
@@ -541,6 +542,127 @@ function mixed_canonize!(tn::AbstractMPO, orthog_center)
     return tn
 end
 
+"""
+    evolve!(ψ::AbstractAnsatz, mpo::AbstractMPO; threshold=nothing, maxdim=nothing, normalize=true, reset_index=true)
+
+Evolve the [`AbstractAnsatz`](@ref) `ψ` with the [`AbstractMPO`](@ref) `mpo` along the output indices of `ψ`.
+If `threshold` or `maxdim` are not `nothing`, the tensors are truncated after each sweep at the proper value, and the
+bond is normalized if `normalize=true`. If `reset_index=true`, the indices of the `ψ` are reset to the original ones.
+"""
+function evolve!(
+    ψ::AbstractAnsatz, mpo::AbstractMPO; threshold=nothing, maxdim=nothing, normalize=true, reset_index=true
+)
+    original_sites = copy(Quantum(ψ).sites)
+    evolve!(form(ψ), ψ, mpo; threshold, maxdim, normalize)
+
+    if reset_index
+        resetindex!(Quantum(ψ); init=ninds(TensorNetwork(ψ)) + 1)
+
+        replacements = [(Quantum(ψ).sites[key] => original_sites[key]) for key in keys(original_sites)]
+        replace!(Quantum(ψ), replacements)
+    end
+
+    return ψ
+end
+
+function evolve!(::NonCanonical, ψ::AbstractAnsatz, mpo::AbstractMPO; threshold, maxdim, normalize, kwargs...)
+    L = nsites(ψ)
+    Tenet.@reindex! outputs(ψ) => inputs(mpo)
+
+    right_inds = [inds(ψ; at=Site(i), dir=:right) for i in 1:(L - 1)]
+
+    for i in 1:L
+        contract_ind = inds(ψ; at=Site(i))
+        push!(ψ, tensors(mpo; at=Site(i)))
+        contract!(ψ, contract_ind)
+        merge!(Quantum(ψ).sites, Dict(Site(i) => inds(mpo; at=Site(i))))
+    end
+
+    # Group the parallel bond indices
+    for i in 1:(L - 1)
+        groupinds!(ψ, right_inds[i])
+    end
+
+    if !isnothing(threshold) || !isnothing(maxdim)
+        truncate_sweep!(form(ψ), ψ; threshold, maxdim, normalize)
+    else
+        normalize && normalize!(ψ)
+    end
+
+    return ψ
+end
+
+function evolve!(::MixedCanonical, ψ::AbstractAnsatz, mpo::AbstractMPO; normalize, kwargs...)
+    initial_form = form(ψ)
+    mixed_canonize!(ψ, Site(nsites(ψ))) # We convert all the tensors to left-canonical form
+
+    evolve!(NonCanonical(), ψ, mpo; normalize, kwargs...)
+
+    mixed_canonize!(ψ, initial_form.orthog_center)
+
+    return ψ
+end
+
+function evolve!(::Canonical, ψ::AbstractAnsatz, mpo::AbstractMPO; threshold, maxdim, normalize, kwargs...)
+    # We first join the λs to the Γs to get MixedCanonical(Site(1)) form
+    for i in 1:(nsites(ψ) - 1)
+        contract!(ψ; between=(Site(i), Site(i + 1)), direction=:right)
+    end
+
+    evolve!(NonCanonical(), ψ, mpo; threshold=nothing, maxdim=nothing, normalize=false, kwargs...) # set maxdim and threshold to nothing so we truncate from Canonical form
+
+    if !isnothing(threshold) || !isnothing(maxdim)
+        truncate_sweep!(Canonical(), ψ; threshold, maxdim, normalize)
+    else
+        normalize && canonize!(ψ; normalize)
+    end
```

Comment:
do you need to normalize if you don't truncate?
also, for this case (i.e. `Canonical` form) you will canonize 2 times if you truncate and normalize (1 in `truncate_sweep` for recanonization and another here for normalization)

```suggestion
    if !isnothing(threshold) || !isnothing(maxdim)
        truncate_sweep!(Canonical(), ψ; threshold, maxdim, normalize)
        normalize && canonize!(ψ; normalize)
    end
```

Reply:
I don't understand this. This is not the case, you don't need to `normalize` if you already normalize the bond that you truncate.

And yes, you might want to normalize if you don't truncate if the `mpo` is not unitary and evolves the `ψ` out of norm.

Reply:
why would you want to normalize here a TN that it's not already normalized? but ok, I see that both things could be understandable from a semantics point of view

let's merge, discuss on Monday and refactor it if we need